initrd/bin/tpmr

```diff
@@ -86,45 +100,179 @@ tpm1_pcrread() {
+                ;;
+        * )     HASH1=$(extend_sha256 "$1" "$2")
+                shift 2
+                extend_sha256 "$HASH1" $@
+                return
+                ;;
+        esac
+        local HASH1_ESC=$(validate_and_escape_hash "$HASH1")
+        local HASH2_ESC=$(validate_and_escape_hash "$HASH2")
+        printf "%b" $HASH1_ESC $HASH2_ESC | sha256sum | sed "s/[[:space:]]*-//"
 }
-tpm1_calcfuturepcr() {
-	TRACE "Under /bin/tpmr:tpm1_calcfuturepcr"
-	if [ "$1" = "-a" ]; then
-		APPEND=y
-		shift
-	fi
 
-	input_file="$1"
-	output_file="$2"
+# There are 3 (and a half) possible formats of event log, each of them requires
+# different arguments for grep. Those formats are shown below as heredocs to
+# keep all the data, including whitespaces:
+# 1) TPM2 log, which can hold multiple hash algorithms at once:
+: << 'EOF'
+TPM2 log:
+	Specification: 2.00
+	Platform class: PC Client
+TPM2 log entry 1:
+	PCR: 2
+	Event type: Action
+	Digests:
+		 SHA256: de73053377e1ae5ba5d2b637a4f5bfaeb410137722f11ef135e7a1be524e3092
+		 SHA1: 27c4f1fa214480c8626397a15981ef3a9323717f
+	Event data: FMAP: FMAP
+EOF
+# 2) TPM1.2 log (aka TCPA), digest is always SHA1:
+: << 'EOF'
+TCPA log:
+	Specification: 1.21
+	Platform class: PC Client
+TCPA log entry 1:
+	PCR: 2
+	Event type: Action
+	Digest: 27c4f1fa214480c8626397a15981ef3a9323717f
+	Event data: FMAP: FMAP
+EOF
+# 3) coreboot-specific format:
+#   3.5) older versions printed 'coreboot TCPA log', even though it isn't TCPA
+: << 'EOF'
+coreboot TPM log:
+
+ PCR-2 27c4f1fa214480c8626397a15981ef3a9323717f SHA1 [FMAP: FMAP]
+EOF
+
+# Returns first word of log - TPM2, TCPA or coreboot
+get_log_format () {
+	echo "$1" | head -n 1 | cut -d " " -f 1
+}
 
-	if [ -z "$APPEND" ]; then
-		true >"$output_file"
+# usage: tpmr calcfuturepcr <pcr_num> [ <input_file>|<input_hash> ... ]
+# Replays PCR value from CBMEM event log. Note that this contains only the
+# measurements performed by firmware, without those performed by Heads (USB
+# modules, LUKS header etc). First argument is PCR number, followed by optional
+# hashes and/or files extended to given PCR after firmware. Resulting PCR value
+# is returned in binary form.
+replay_sha1_pcr () {
+	TRACE "Under /bin/tpmr:replay_sha1_pcr"
+	if [ -z "$1" ] ; then
+		>&2 echo "No PCR number passed"
+		return
 	fi
+	if [ "$1" -ge 8 ] ; then
+		>&2 echo "Illegal PCR number ($1)"
+		return
+	fi
+	local LOG=`cbmem -L`
+	local PCR='PCR: '
+	local LINES=0
+	local ALG='SHA1'
+	case `get_log_format "$LOG"` in
+	"coreboot" )	FMT='PCR-'
+			;;
+	"TCPA" )	LINES=2
+			ALG='Digest:'
+			;;
+	"TPM2" )	# With 1 hash algo we would get digest of next entry on
+			# 9th line after 'PCR: '. Read as much as possible in
+			# case TPM implements high number of algorithms.
+			LINES=8
+			;;
+	* )		>&2 echo "Unknown TPM event log format"
+			return
+			;;
+	esac
+	PCR="${PCR}$1"
+	shift
+	extend_sha1 $sha1_zeros \
+	`echo "$LOG" | grep "$PCR" -A$LINES | grep $ALG | cut -d " " -f 3` \
```

Comment:
Since this is trying to carry state from line to line, awk would be a better fit than multiple-grep.

The number of 'after' lines is fragile because it must be long enough to ensure that the $ALG line for that entry is found, but not so long that it accidentally picks up $ALG for the _next_ entry (which might be a different PCR).  This seems likely to break in subtle ways if the format changes in coreboot, either due to added or removed lines.

With awk you could set a variable when finding a matching PCR line, clear it when finding a non-matching PCR line, and emit the $ALG match only when we are in a section for the correct PCR.

Also, I suggest `$(...)` over backticks: https://www.shellcheck.net/wiki/SC2006

Reply:
Unfortunately, whole format of log printed by coreboot is fragile, and during upstreaming effort we encountered some resistance from the maintainers that initially didn't seem to understand the importance of following standards. Ideally all formats should have constant, well-documented, machine-parsable format, but we can't do anything about "old" coreboot format, and there still are (and probably will be for some time, because nobody cares) boards that use it.

I'll change this to use awk, for some reason I thought that it wasn't available here.

> Also, I suggest $(...) over backticks

Right, I had it that way originally but changed to backticks to debug missing `"` that was in fact missing `;;` and forgot to change it back.